surefire-booter/src/test/java/org/apache/maven/surefire/booter/SystemUtilsTest.java

```diff
@@ -98,6 +98,15 @@ public void incorrectJdkPath()
             assertThat( SystemUtils.isJava9AtLeast( incorrect.getAbsolutePath() ) ).isFalse();
         }
 
+        @Test
+        public void incorrectJdkPathShouldNotNPE()
```

Comment:
is `jvmExecutable` set to `/opt/jdk` instead of `/opt/jdk/bin/java` in your case? this test is written to make `jdk=/opt` and `jre=/opt/jdk/` which I can understand you want a case with a single segment path but is not aligned on the description and only possible if the jdk is extracted at the root of the filesystem, is it your case? Also it means the jvmExecutable is not the executable so something is wrong higher in the processing and must be fixed preventing this NPE later so I would rather chase this cause instead of swallowing it.

Any inputs on these points?

Reply:
Well, the reason I submitted this is because unit tests fail, due to this: on my machine JAVA_HOME is `/opt/openjdk-bin-11/`. Unit test `incorrectJdkPath()` works as follows:

```
        @Test
        public void incorrectJdkPath()
        {
            File jre = new File( System.getProperty( "java.home" ) );
            File jdk = jre.getParentFile();
            File incorrect = jdk.getParentFile();
            assertThat( SystemUtils.isJava9AtLeast( incorrect.getAbsolutePath() ) ).isFalse();
        }
```

This leads to `incorrect=/`, which reaches `toJdkHomeFromJvmExec("/")`, which calls again `getParent("/")`, which yields null, and next `getName()` is called on this null. It is clear from the implementation of `toJdkHomeFromJvmExec()` that returning null is fine when this jvmExec is not found. I believe this is the case here.

Also, the test is about incorrect JDK paths, and such a path is an incorrect JDK path. However, one incorrect case crashes with a non-informative NPE. I would say a "cannot find jvm executable" error on the return-null path would be more helpful for users that misconfigured their systems.

Reply:
@CMoH
We can accept this fix for the unit test but more important would be to back track the callers in `AbstractSurefireMojo` and see if this NPE may happen and may have some influence, and where if any.

Reply:
I've backtracked the production scope callers of the SystemUtils method and the parameter can come from three possible places:

1. `<jvm>` mojo property
2. `System.getProperty( "java.home" ) + File.separator + "bin" + File.separator + "java"`
3. JavaToolChain installFolder `+ File.separator + "bin" + File.separator + "java"`

The only one that could in theory lead to the NPE is the `<jvm>` property - but only if a user enters `/` or `C:\` there. Not an actual use case.

Conclusion: This is not an actual problem in production.